### ISSUE TITLE
Nightly test failure openmp.scatterview: values diff

### ISSUE THREAD

**Issue by ndellingwood** [kokkos/kokkos#5025]
**Describe the bug**
Following a merge on 5/16/2022 the `openmp.scatterview` is failing with gcc/6.4.0 builds (so far):

```
18:25:21 [ RUN      ] openmp.scatterview
18:25:21 /home/jenkins/weaver-new/workspace/Kokkos_develop_weaver_gcc_640/kokkos/containers/unit_tests/TestScatterView.hpp:365: Failure
18:25:21 The difference between val2 and 1.0 is 1, which exceeds 1e-14 * 1.0, where
18:25:21 val2 evaluates to 2,
18:25:21 1.0 evaluates to 1, and
18:25:21 1e-14 * 1.0 evaluates to 1e-14.
18:25:21 Data differs at index 5
18:25:21 [  FAILED  ] openmp.scatterview (223423 ms)
```

**Reproducer**:
```
module load cmake/3.19.3 gcc/6.4.0

$KOKKOS_PATH/generate_makefile.bash --with-devices=OpenMP --arch=Power9 --compiler=g++ --cxxflags="-O3 -Wall -Wunused-parameter -Wshadow -pedantic -Werror -Wsign-compare -Wtype-limits -Wignored-qualifiers -Wempty-body -Wclobbered -Wuninitialized " --cxxstandard="14" --kokkos-path=$KOKKOS_PATH  --no-examples --with-options=enable_large_mem_tests
```

**Comment by ndellingwood** [comment]
The same build passed tonight, so the failure appears to be spurious

**Comment by PhilMiller** [comment]
I would take it as a sign of a race condition. Especially given that it's Power, with weaker memory ordering guarantees than x86.

We should run the test under Thread Sanitizer, and actually fix the hazards it reports

**Comment by ndellingwood** [comment]
More failures on 5/17/2022 of `openmp.scatterview`, gcc/9.2.0 build and clang/7.0+cuda/9.2.0 nightly tests

**Comment by dalg24** [comment]
@ndellingwood please confirm that all these failures were `openmp.scatterview`

**Comment by ndellingwood** [comment]
@dalg24 correct, all failures were with openmp.scatterview, matching those in the original post

**Comment by PhilMiller** [comment]
I built with `CMAKE_CXX_FLAGS=-fsanitize=thread`, and am running

> `OMP_NUM_THREADS=2 TSAN_OPTIONS="log_path=tsanlog suppressions=${HOME}/repos/kokkos/tsan-suppressions" ./containers/unit_tests/KokkosContainers_UnitTest_OpenMP --gtest_filter=openmp.scatterview`

I have the following suppressions in that file:
```
race:OpenMPSpaceInitializer
race_top:Kokkos_OpenMP_Parallel.hpp
race:Test::test_scatter_view_impl_cls<Kokkos::OpenMP, Kokkos::LayoutRight, Kokkos::Experimental::ScatterNonDuplicated, Kokkos::Experimental::ScatterAtomic, Kokkos::Experimental::ScatterSum, int>::initialize
race_top:ViewFill
```

I'm still getting 89 reports from Thread Sanitizer.

As I read them, I'm somewhat concerned that our OpenMP backend is now actually behaving incorrectly, and potentially running some iterations of the `parallel_for` repeatedly or not at all.

**Comment by ndellingwood** [comment]
Nightly failure 05/18/2022 of `openmp.scatterview` with gcc/8.3.0 on x86

**Comment by PhilMiller** [comment]
On Vortex (Power9), built with `gcc/8.3.1`, I can build as follows
```
cmake ~/repos/kokkos  -DCMAKE_CXX_FLAGS=-DKOKKOS_INTERNAL_DISABLE_NATIVE_OPENMP -DCMAKE_BUILD_TYPE=RelWithDebInfo -DKokkos_ENABLE_OPENMP=yes -DKokkos_ENABLE_TESTS=yes
```
And then test as follows:
```
GTEST_FILTER=openmp.scatterview ctest --repeat-until-fail 100 -R KokkosContainers_UnitTest_OpenMP
```
With no failures

**Comment by PhilMiller** [comment]
Huh, I'm re-running that test binary in full with the newer OpenMP code, and I'm now not seeing the failures reproduce

**Comment by PhilMiller** [comment]
Failures seen in 1000 runs of `OMP_NUM_THREADS=N GTEST_FILTER=openmp.scatterview_devicetype ctest -R KokkosContainers_UnitTest_OpenMP`

| Threads | Old Code | New Code |
| --- | --- | --- | 
| 4 | 0 | 4 |
| 8 | 0 | 132 |
| 16 | 80 | 76 |
| 24 | 60 | 74 |

**Comment by PhilMiller** [comment]
For future reference, since we were trying to use Thread Sanitizer and it's super noisy on GNU's libgomp, here's links to issues of OpenMP support:

* https://gcc.gnu.org/bugzilla/show_bug.cgi?id=55561 - idle since 2014
* https://reviews.llvm.org/D13072 - added to LLVM's libomp in 2016

**Comment by dalg24** [comment]
Fix in develop and cherry-picked into 3.6.1 release candidate branch
